File: javascript/protein-translation/protein-translation.js
//
// This is only a SKELETON file for the 'Protein Translation' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

// I did it without lodash. The difficulty of the exercise does not look like 'Easy'.

const codonProteinTable = {
  'AUG': 'Methionine',
  'UUU': 'Phenylalanine',
  'UUC': 'Phenylalanine',
  'UUA': 'Leucine',
  'UUG': 'Leucine',
  'UCU': 'Serine',
  'UCC': 'Serine',
  'UCA': 'Serine',
  'UCG': 'Serine',
  'UAU': 'Tyrosine',
  'UAC': 'Tyrosine',
  'UGU': 'Cysteine',
  'UGC': 'Cysteine',
  'UGG': 'Tryptophan',
};

const goodCodons = Object.keys(codonProteinTable);
const stopCodons = ['UAA', 'UAG', 'UGA'];

const chunkString = (string, base) => {
  const iter = (acc, str) => {
    if (str === '') {
      return acc;
    }

    return iter([...acc, str.slice(0, base)], str.slice(base));
  };

  return iter([], string);
};

export const translate = (strand) => {
  if (!strand) return [];

  const codons = chunkString(strand, 3);
  const proteins = codons.reduce((acc, codon) => {
    if (stopCodons.includes(codon)) {
      return [...acc, 'STOP'];

    } else if (!goodCodons.includes(codon)) {
      throw new Error("Invalid codon");
    }
    return [...acc, codonProteinTable[codon]];
  }, []);
  const stopIndex = proteins.indexOf('STOP');

  return proteins.includes('STOP') ? proteins.slice(0, stopIndex) : proteins;
};

File: ruby/protein-translation/protein_translation.rb
# frozen_string_literal: true

# Write your code for the 'Protein Translation' exercise in this file. Make the tests in
# `protein_translation_test.rb` pass.
#
# To get started with TDD, see the `README.md` file in your
# `ruby/protein-translation` directory.

class InvalidCodonError < StandardError
  def initialize
    super
  end
end

class Translation
  CODON_PROTEIN_TABLE = {
    'AUG' => 'Methionine',
    'UUU' => 'Phenylalanine',
    'UUC' => 'Phenylalanine',
    'UUA' => 'Leucine',
    'UUG' => 'Leucine',
    'UCU' => 'Serine',
    'UCC' => 'Serine',
    'UCA' => 'Serine',
    'UCG' => 'Serine',
    'UAU' => 'Tyrosine',
    'UAC' => 'Tyrosine',
    'UGU' => 'Cysteine',
    'UGC' => 'Cysteine',
    'UGG' => 'Tryptophan'
  }.freeze
  STOP_CODONS = %w[UAA UAG UGA].freeze
  VALID_CODONS = CODON_PROTEIN_TABLE.keys.concat(STOP_CODONS).freeze

  class << self
    def of_codon(codon)
      raise InvalidCodonError unless VALID_CODONS.include?(codon)

      STOP_CODONS.include?(codon) ? 'STOP' : CODON_PROTEIN_TABLE[codon]
    end

    def of_rna(strand)
      codons = strand.scan(/.{1,3}/)
      return 'STOP' if STOP_CODONS.include?(codons.first)

      proteins = codons.map { |c| of_codon(c) }
      stop_index = proteins.index('STOP').nil? ? -1 : proteins.index('STOP') - 1
      proteins.slice(0..stop_index)
    end
  end
end
